aten/src/ATen/FunctionalTensorImplBase.cpp

```diff
@@ -0,0 +1,184 @@
+      keyset.add(DispatchKey::Functionalize),
+      dtype,
+      device
+    ) {}
+
+FunctionalTensorImplBase::FunctionalTensorImplBase(caffe2::TypeMeta dtype, c10::Device device)
+  : TensorImpl(
+      c10::DispatchKeySet(DispatchKey::Functionalize),
+      dtype,
+      device
+    ) {}
+
+void FunctionalTensorImplBase::mark_as_alias() {
+  is_alias_tensor_ = true;
+}
+
+void FunctionalTensorImplBase::sync_() {
+  if (is_up_to_date()) {
+    return;
+  }
+  // Apply all updates on alias_
+  alias_->SyncUpdateOperations();
+  // Reapply views to Get the viewed tensor from updated base in alias_
+  auto t = alias_->base();
+  for (auto& view_meta: view_metas_) {
+    switch (view_meta.view_type) {
+      case ViewMeta::Type::view:
+          t = t.view(view_meta.size);
+          break;
+      case ViewMeta::Type::noOp:
+          break;
+      default:
+          TORCH_CHECK(false, "Tried to run the functionalization pass on an unsupported view: ", view_meta.view_type);
+    }
+  }
+  // We want the new tensor to have separate memory from the alias.
+  // Note that the clone required for functorch, but probably unnecessary for backends
+  // like vulkan and xla, because their view operators already create fresh tensors.
+  t = t.clone();
+  // This call to replace_() doesn't go through the dispatcher.
+  // It's a virtual method implemented directly on TensorImpl subclasses.
+  replace_(t.unsafeGetTensorImpl());
+  generation_ = alias_->generation();
+}
+
+bool FunctionalTensorImplBase::is_up_to_date() const {
+  if (alias_) {
+    return generation_ == alias_->generation();
+  }
+  return true;
+}
+
+void FunctionalTensorImplBase::maybe_add_update(const Tensor& updated_val) {
+  // If the mutated tensor wasn't a view, we don't need to do anything
+  if (is_view()) {
+    alias_->add_update(updated_val.clone(), view_metas_);
+    // TODO: add this back see if anything breaks?
+    //generation_ = alias_->generation(); // self is fully up to date at this point
+  }
+}
+
+void FunctionalTensorImplBase::set_view_meta(const Tensor& other, at::ViewMeta meta) {
+    auto other_impl = dynamic_cast<at::FunctionalTensorImplBase*>(other.unsafeGetTensorImpl());
+    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(other_impl != nullptr);
+    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(alias_ == nullptr);
+    if (other_impl->alias_ != nullptr) {
+        // if the other tensor is already a view, copy its ViewMeta vector and push the current one.
+        std::vector<at::ViewMeta> metas = other_impl->view_metas_;  // copy
+        metas.push_back(meta);
+        //alias_ = other_impl->alias_; // refcount bump
+        alias_ = other_impl->alias_; // refcount bump
+        view_metas_ = metas;
+    } else {
+        std::shared_ptr<Alias> alias = std::make_shared<Alias>(const_cast<Tensor&>(other));
+        at::ViewMeta noop_view_info(at::ViewMeta::Type::noOp, other.sizes().vec(), other.sizes().vec());
+        // The original tensor wasn't a view - turn it into a (no-op) view.
```

Comment:
Warning: This isn't thread safe; if multiple threads try to take out a view on the same non-view tensor, you can end up with two aliases (one of which is disconnected). Easiest way to fix this is probably to make `alias_` an atomic shared pointer (except we're not C++20 so you can't actually do that). Not sure what the actual easiest fix here is: we can probably ignore it for now but then we need to prominently document that functionalization is not thread safe.

Reply:
oof, yeah. is multithreading in the presence of views/mutations on xla a use case people care about (with different threads looking at different views)? If it is, maybe we should just add locking logic here. Otherwise documenting it somewhere seems reasonable

Reply:
Oh, well I don't see anything indicating that xla's current alias creation is thread safe. So at least we're not making things worse (for xla, anyway). xla's alias creation is here: https://github.com/pytorch/xla/blob/master/torch_xla/csrc/tensor.cpp#L804